geo/Geoserver.py

```diff
@@ -430,23 +432,55 @@ def upsert_workspaces_rules(self, workspacePattern: str = '*', permission: str =
 
         created = 0
         updated = 0
+        deleted = 0
+
+        rulesList = self.get_layer_rules()
 
         for workspace in workspaceList:
             try:
-                self.create_layer_rule(workspace['name'], '*', permission, role)
-                created += 1
+                if mode != "D":
+                    self.create_layer_rule(workspace['name'], '*', permission, role)
+                    created += 1
+                else:
+                    error_data = {
+                        'status_code': 409,
+                    }
+                    raise Exception(
+                        "It is not possible to create a rule in Delete mode", error_data)
             except Exception as create_exception:
-                try:
-                    message, error_data = create_exception.args
-                    status_code = error_data['status_code']
-                    if status_code == 409:
-                        self.update_layer_rule(workspace['name'], '*', permission, role)
+                message, error_data = create_exception.args
+                status_code = error_data['status_code']
+                rule = workspace['name'] + '.' + '*' + '.' + permission
+                prevRole = rulesList.get(rule, None)
+                newRole = role
+
+                if prevRole is None:
```

Comment:
If I understand correctly, this is the case where the rule should be created because it doesn't exist, but it couldn't be created, meaning `create_layer_rule` raised "The rule can not be created". In that case, maybe we should add another counter for rules that couldn't be created/updated ?

Reply:
In general, when there is an exception trying to create a rule (inside the `try` block) and the current rule (`workspace.*.permission`) does not exist in the list of defined rules, the `prevRol` value is set to `None`.  

For example, if the user executes `geo.upsert_workspaces_rules('taxid-000', 'r', 'AAAA', 'D')` and there are no rules defined for the workspace `taxid-000`, in this case, `prevRol` is `None`.  

Initially, I was not convinced to add a new counter, but now I think it is a good idea to let the user know that the rules were not created, although the program attempted to do so and encountered some failures. even, i added some other cases to `failures` counter.